### ISSUE TITLE
decorator based model serving

### ISSUE THREAD

**Issue by msaroufim** [pytorch/serve#1458]
I'd like us to move to an experience like this when serving models with torchserve to follow projects like Fast API

```python
@serve(endpoint, model_path)
def inference():
  inferencing_logic...

if __name__ == __main__:
 inference()
```

When `inference()` is called it would go ahead and start torchserve by automatically creating a mar file and generating a config property with the required endpoint and load the default base handler.

So perhaps `serve()` would be implemented as 

```python
def serve(endpoint, model, path):
 mar_file = create_mar_file(model_path)
 config_properties = create_config(endpoint)
 os.system("torchserve --start --model_path mar_file --config config.properties 
```

**Comment by suraj813** [comment]
I love the idea of reducing shell boilerplate! How does this play with custom handlers? 

In this new paradigm, instead of defining a new `CustomHandler(BaseHandler)` subclass, can we simply run 
```
@serve(endpoint, model_path)
def handle():
 preprocess()
 inference()
 postprocess()
```
in a python file and expect the decorated logic to run in lieu of the `base_handler`? IIUC `torchserve --start` accesses the handler file from the model archiver?

**Comment by msaroufim** [comment]
> I love the idea of reducing shell boilerplate! How does this play with custom handlers?
> 
> In this new paradigm, instead of defining a new `CustomHandler(BaseHandler)` subclass, can we simply run
> 
> ```
> @serve(endpoint, model_path)
> def handle():
>  preprocess()
>  inference()
>  postprocess()
> ```
> 
> in a python file and expect the decorated logic to run in lieu of the `base_handler`?

Yup that's the intent, It's like setting up a handler by writing a main function

**Comment by mreso** [comment]
Love the idea too, will this be limited to a single model? How would we handle a case where a torchserve process is already running (Fail? Add model to running torchserve?)?

**Comment by msaroufim** [comment]
Thanks for the feedback @mreso 
* Single model: interesting callout, in principle there is no hard limitation just can't trivially call `torchserve --start` multiple times and make sure the same `model_store` is used
* Failures: If an inference fails, right now we just say something like worker died - #1475 is making this a bit better so we can surface errors in console instead of asking users to navigate to `logs/model_log.log` which TBH is very confusing
* Add models: I think the general question here is how do we work with the management API in this regime, since management calls are synchronous I think this can be as simple as wrapping `requests` library with some more limited arguments that we support
